implementations/rust/ockam/ockam_transport_tcp/src/transport/lifecycle.rs

```diff
@@ -31,3 +43,138 @@ impl TcpTransport {
         &self.registry
     }
 }
+
+#[async_trait]
+impl Transport for TcpTransport {
+    fn transport_type(&self) -> TransportType {
+        TCP
+    }
+
+    async fn resolve_route(&self, flow_controls: &FlowControls, route: Route) -> Result<Route> {
+        let mut result = Route::new();
+        let mut number_of_tcp_hops = 0;
+
+        for address in route.iter() {
+            if address.transport_type() == TCP {
+                if number_of_tcp_hops >= 1 {
+                    return Err(Error::new(
+                        Origin::Transport,
+                        Kind::Invalid,
+                        "only one TCP hop is allowed in a route",
+                    ));
+                }
+
+                let options = if SocketAddr::from_str(address.address())
+                    .map(|socket_addr| socket_addr.ip().is_loopback())
+                    .is_ok()
+                {
+                    // TODO: Enable FlowControl for loopback addresses as well
+                    TcpConnectionOptions::insecure()
+                } else {
+                    let id = flow_controls.generate_id();
+                    TcpConnectionOptions::as_producer(flow_controls, &id)
+                };
+
+                number_of_tcp_hops += 1;
+                let addr = self.connect(address.address().to_string(), options).await?;
+                result = result.append(addr)
+            } else {
+                result = result.append(address.clone());
+            }
+        }
+
+        let resolved = result.into();
+        debug!("resolved route {} to {}", route, resolved);
+        Ok(resolved)
```

Comment:
Should we close the opened connection if this function returns the Error?

Reply:
That's a good suggestion indeed. However when I try that the second unit test (checking the number of hops) hangs and I can't see why yet. Any idea?

Reply:
I fixed the issue by simply checking the number of hops first and only creating a connection if the route is correct.